pkg/chain/ethereum/ethutil/rate_limiter.go

```diff
@@ -0,0 +1,218 @@
+package ethutil
+
+import (
+	"context"
+	"fmt"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"golang.org/x/sync/semaphore"
+	"golang.org/x/time/rate"
+	"math/big"
+	"time"
+)
+
+type rateLimiter struct {
+	bind.ContractBackend
+
+	limiter   *rate.Limiter
+	semaphore *semaphore.Weighted
+
+	acquirePermitTimeout time.Duration
+}
+
+// RateLimiterConfig represents the configuration of the rate limiter.
+type RateLimiterConfig struct {
+	// RequestsPerSecondLimit sets the maximum average number of requests
+	// per second. It's important to note that in short periods of time
+	// the actual average may exceed this limit slightly.
+	RequestsPerSecondLimit int
+
+	// ConcurrencyLimit sets the maximum number of concurrent requests which
+	// can be executed against the underlying contract backend in the same time.
+	ConcurrencyLimit int
+
+	// AcquirePermitTimeout determines how long a request can wait trying
+	// to acquire a permit from the rate limiter.
+	AcquirePermitTimeout time.Duration
+}
+
+// WrapRateLimiting wraps the given contract backend with rate limiting
+// capabilities with respect to the provided configuration.
+func WrapRateLimiting(
+	backend bind.ContractBackend,
+	config *RateLimiterConfig,
+) bind.ContractBackend {
+	rateLimiter := &rateLimiter{ContractBackend: backend}
+
+	if config.RequestsPerSecondLimit > 0 {
+		rateLimiter.limiter = rate.NewLimiter(
+			rate.Limit(config.RequestsPerSecondLimit),
+			1,
+		)
+	}
+
+	if config.ConcurrencyLimit > 0 {
+		rateLimiter.semaphore = semaphore.NewWeighted(
+			int64(config.ConcurrencyLimit),
+		)
+	}
+
+	if config.AcquirePermitTimeout > 0 {
```

Comment:
Shouldn't we set this value only when `RequestsPerSecondLimit` is set? `rateLimiter.acquirePemitTimeout` does not make sense without `rateLimiter.limiter`. What do you think about moving `if config.AcquirePermitTimeout > 0` into `if config.RequestsPerSecondLimit > 0` for clarity?

```
if config.RequestsPerSecondLimit > 0 {
    rateLimiter.limiter = // ...

   if config.AcquirePermitTimeout > 0 {
       // ...
    }
}
```

Reply:
That timeout is also used by the `rateLimiter.semaphore`

Reply:
Ah, right, I missed it. `rl.semaphore.Acquire` is accepting `ctx` that has been created with `rl.acquirePermitTimeout`.